src/software/robot_diagnostics/widgets/drive.cpp

```diff
@@ -0,0 +1,113 @@
+#include "software/robot_diagnostics/widgets/drive.h"
+
+void setupDrive(Ui::AutoGeneratedMainWidget* widget,
+                std::function<void(double, DirectPerWheelMode)>
+                    direct_per_wheel_power_changed_callback,
+                std::function<void(double, DirectVelocityMode)>
+                    direct_velocity_power_changed_callback)
+{
+    setupDirectPerWheel(widget, direct_per_wheel_power_changed_callback);
+    setupDirectVelocity(widget, direct_velocity_power_changed_callback);
+}
+
+
+void setupDirectPerWheel(Ui::AutoGeneratedMainWidget* widget,
+                         std::function<void(double, DirectPerWheelMode)>
+                             direct_per_wheel_power_changed_callback)
+{
+    auto direct_per_wheel_power_fl_changed = [direct_per_wheel_power_changed_callback](
+                                                 double value) {
+        direct_per_wheel_power_changed_callback(value, DirectPerWheelMode::FRONT_LEFT);
+    };
+    auto direct_per_wheel_power_fr_changed = [direct_per_wheel_power_changed_callback](
+                                                 double value) {
+        direct_per_wheel_power_changed_callback(value, DirectPerWheelMode::FRONT_RIGHT);
+    };
+    auto direct_per_wheel_power_bl_changed =
+        [direct_per_wheel_power_changed_callback](double value) {
+            direct_per_wheel_power_changed_callback(value, DirectPerWheelMode::BACK_LEFT);
+        };
+    auto direct_per_wheel_power_br_changed = [direct_per_wheel_power_changed_callback](
+                                                 double value) {
+        direct_per_wheel_power_changed_callback(value, DirectPerWheelMode::BACK_RIGHT);
+    };
+    setupSliderLineEdit(widget->lineEdit_direct_per_wheel_fl,
+                        widget->slider_direct_per_wheel_fl,
+                        direct_per_wheel_power_fl_changed, -100.0, 100.0, 1.0);
```

Comment:
what are the units here, is this a %? Im not even sure what it is on the old MRF test

Reply:
some measure of power and we can think of it like a % I think the idea is that it's a bit of work to calibrate for rpm or something more meaningful but we can revisit this

Reply:
Yeah agreed. Could we actually make a note in the GUI here like `No Units` to be super explicit?

IIRC this used to effectively be PWM, then I changed it to centi-newtons, but that's too small so it's probably getting changed again.